src/_pytest/assertion/util.py

```diff
@@ -439,7 +442,8 @@ def _compare_eq_cls(
         explanation += ["Differing attributes:"]
         for field in diff:
             explanation += [
-                ("%s: %r != %r") % (field, getattr(left, field), getattr(right, field))
+                ("%s: %r != %r") % (field, getattr(left, field), getattr(right, field)),
+                *_compare_eq_any(getattr(left, field), getattr(right, field), verbose),
```

Comment:
```suggestion 
                "",  
                "Recursive from previous comparison",
                *_compare_eq_any(getattr(left, field), getattr(right, field), verbose),
```

I've tested the PR locally, and one blank line and a short explanation between each recursion block helps readability, IMO:

```
>       assert left == right
E       AssertionError: assert ComplexDataObject(field_a=SimpleDataObject(field_a=1, field_b='b'), field_b=ComplexDataObject2(field_a=SimpleDataObject(field_a=1, field_b='b'), field_b=SimpleDataObject(field_a=2, field_b='c'))) == ComplexDataObject(field_a=SimpleDataObject(field_a=1, field_b='b'), field_b=ComplexDataObject2(field_a=SimpleDataObject(field_a=1, field_b='b'), field_b=SimpleDataObject(field_a=3, field_b='c')))
E         Matching attributes:
E         ['field_a']
E         Differing attributes:
E         field_b: ComplexDataObject2(field_a=SimpleDataObject(field_a=1, field_b='b'), field_b=SimpleDataObject(field_a=2, field_b='c')) != ComplexDataObject2(field_a=SimpleDataObject(field_a=1, field_b='b'), field_b=SimpleDataObject(field_a=3, field_b='c'))
E
E         Recursive from previous comparison
E         Matching attributes:
E         ['field_a']
E         Differing attributes:
E         field_b: SimpleDataObject(field_a=2, field_b='c') != SimpleDataObject(field_a=3, field_b='c')
E
E         Recursive from previous comparison
E         Matching attributes:
E         ['field_b']
E         Differing attributes:
E         field_a: 2 != 3
E
E         Recursive from previous comparison
E         +2
E         -3
```